packages/catalog-generator/src/main/java/io/kaoto/camelcatalog/maven/KaotoMavenVersionManager.java

```diff
@@ -0,0 +1,177 @@
+    private boolean log;
+
+    private KaotoMavenVersionManager(MavenDownloaderImpl downloader) {
+        this.downloader = downloader;
+        downloader.build();
+    }
+
+    public KaotoMavenVersionManager() {
+        this(new MavenDownloaderImpl());
+        this.setClassLoader(new KaotoOpenURLClassLoader());
+    }
+
+    public boolean getLog() {
+        return log;
+    }
+
+    public void setLog(boolean log) {
+        this.log = log;
+    }
+
+    /**
+     * To add a 3rd party Maven repository.
+     *
+     * @param name the repository name
+     * @param url  the repository url
+     */
+    public void addMavenRepository(String name, String url) {
+        super.addMavenRepository(name, url);
+        repositories.put(name, url);
+    }
+
+    @Override
+    public String getLoadedVersion() {
+        return version;
+    }
+
+    @Override
+    public String getRuntimeProviderLoadedVersion() {
+        return runtimeProviderVersion;
+    }
+
+    @Override
+    public boolean loadRuntimeProviderVersion(String groupId, String artifactId, String version) {
+        try {
+            MavenDownloader mavenDownloader = downloader;
+            String gav = String.format("%s:%s:%s", groupId, artifactId, version);
+
+            boolean useSnapshots = version.endsWith("SNAPSHOT");
+            resolve(mavenDownloader, gav, useSnapshots, false);
+
+            if (artifactId.contains("catalog")) {
+                this.version = version;
+            } else {
+                this.runtimeProviderVersion = version;
+            }
+
+            return true;
+        } catch (Exception e) {
+            if (getLog()) {
+                LOGGER.log(Level.WARNING,
+                        String.format("Cannot load runtime provider version {} due {}", version, e.getMessage()), e);
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Resolves Maven artifact using passed coordinates and use downloaded artifact
+     * as one of the URLs in the
+     * helperClassLoader, so further Catalog access may load resources from it.
+     */
+    public void resolve(MavenDownloader mavenDownloader, String gav, boolean useSnapshots, boolean transitive) {
+        try {
+            Set<String> extraRepositories = new LinkedHashSet<>(repositories.values());
+
+            List<MavenArtifact> artifacts = mavenDownloader.resolveArtifacts(Collections.singletonList(gav),
+                    extraRepositories, transitive, useSnapshots);
+
+            if (getLog()) {
+                LOGGER.log(Level.INFO, () -> "Artifacts: " + artifacts);
+            }
+
+            for (MavenArtifact ma : artifacts) {
+                ((KaotoOpenURLClassLoader) getClassLoader()).addURL(ma.getFile().toURI().toURL());
+            }
+        } catch (Throwable e) {
+            if (getLog()) {
+                LOGGER.log(Level.WARNING,
+                        String.format("Cannot resolve artifact {} due to {}", gav, e.getMessage()), e);
```

Comment:
here it is assuming that it is an error resolving the artifact, it can be another kind of error